[Distributed] add pack-check method for float8_e5m2

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136115
* #135961
* #135891

Add support for Float8_e5m2, following similar algorithm used for Float8_e4m3fn (i.e. overflow check).

Made `HasNanFP8x8` a template so that it is extendable based on dtype.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[Distributed] add pack-check method for float8_e5m2

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/NanCheck.cu:
```diff
@@ -79,32 +79,66 @@ struct CheckBytePack<T, /*EltPerPack*/8> {
+// We want to check 8 x FP8 simultaneously, hence this template definition.
+template<typename T>
+struct HasNanFP8x8 {
+  // I am a dumb implementation. You should never call in here, unless the check
```
Static Assert False to raise a compile error if we call in here?

Skylion007 replied:
Or at least can we issue warning somehow?

kwen2501 replied:
I struggle between providing basic functionality so that user code can run without break (current code) vs speed. And eventually chose the former :)

kwen2501 replied:
But your point seems better: compile error can force a developer to implement the template if they want to add a new data type to `AT_DISPATCH_FLOATING_TYPES_AND4` below.

kwen2501 replied:
Hmm, it seems `static_assert` doesn't work well with template prior to c++23. I may try `=delete` as suggested [here](https://stackoverflow.com/questions/44059557/whats-the-right-way-to-call-static-assertfalse).

Skylion007 replied:
@kwen2501 You could also at least unroll the loop, but the speed gains would be minimal.

Skylion007 replied:
You could also use the `self != self ` to check for NaN based on the dtype, but I am not sure if that's faster / more portable.

kwen2501 replied:
> @kwen2501 You could also at least unroll the loop, but the speed gains would be minimal unless the compiler realizes to inline isnan

We tried that. Since the final result is a reduction, i.e.
packHasNan = isnan(byte0) || isnan(byte1) ... || isnan(byte7),
the compiler does not seem quite willing the unroll the loop.

Skylion007 replied:
@kwen2501 It doesn't like `#pragma unroll`?

Skylion007 replied:
We could also manually unroll since it's only 8 elements (as painful as that would be).

Skylion007 commented on torch/csrc/distributed/c10d/NanCheck.cu:
```diff
@@ -79,32 +79,66 @@ struct CheckBytePack<T, /*EltPerPack*/8> {
+template<>
+struct HasNanFP8x8<c10::Float8_e5m2> {
+  static __device__ __forceinline__ bool check(uint64_t fp8x8) {
+    auto t = fp8x8 & 0x7F7F7F7F7F7F7F7FULL;
```
Interesting, we can't use std::numerics_limit for either of these FP8 to check NaN equality? Or is this just faster?

Skylion007 replied:
Ah wait, I see it's packed fp8, nvm.

kwen2501 replied:
Yeah, speed is one of the reasons -- I don't know how to teach CUDA to apply `isnan()` (or any mapping) to each byte of `fp8x8`, followed by a reduction of the results, and still run fast.

Please let me know if you have any suggestion!

Skylion007 replied:
@kwen2501 You could use the self comparison value trick to check if NaN, but I'm not sure if that's faster.

kwen2501 committed:
Update on "[Distributed] add pack-check method for float8_e5m2"


Add support for Float8_e5m2, following similar algorithm used for Float8_e4m3fn (i.e. overflow check).

Made `HasNanFP8x8` a template so that it is extendable based on dtype.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
